extensions/amp-date-picker/1.0/component/date-range-picker.tsx

```diff
@@ -0,0 +1,432 @@
+import {useDatePickerState} from './use-date-picker-state';
+
+import {
+  DateFieldType,
+  DateRangePickerAPI,
+  DateRangePickerProps,
+} from '../types';
+
+function DateRangePickerWithRef(
+  {
+    allowBlockedEndDate,
+    allowBlockedRanges,
+    children,
+    endInputSelector,
+    initialVisibleMonth,
+    mode,
+    openAfterClear,
+    openAfterSelect,
+    startInputSelector,
+    today,
+  }: DateRangePickerProps,
+  ref: Ref<DateRangePickerAPI>
+) {
+  const containerRef = useRef<HTMLElement>(null);
+
+  const [focusedInput, setFocusedInput] =
+    useState<DateFieldType>('start-input');
+
+  const defaultMonth = initialVisibleMonth || today;
+  const [month, setMonth] = useState<Date>(defaultMonth);
+
+  const {isOpen, transitionTo} = useDatePickerState(mode);
+
+  const {
+    blockedDates,
+    formatDate,
+    getDisabledAfter,
+    getDisabledBefore,
+    isDisabled,
+    parseDate,
+  } = useDatePickerContext();
+
+  const startDateInput = useDatePickerInput({
+    inputSelector: startInputSelector,
+    type: 'start-input',
+  });
+
+  const endDateInput = useDatePickerInput({
+    inputSelector: endInputSelector,
+    type: 'end-input',
+  });
+
+  /**
+   * Sets the selected date, month, and input value
+   */
+  const handleSetStartDate = useCallback(
+    (date?: Date) => {
+      if (!date) {
+        return;
+      }
+      startDateInput.handleSetDate(date);
+      setMonth(date);
+    },
+    [startDateInput]
+  );
+
+  /**
+   * Sets the selected date, month, and input value
+   */
+  const handleSetEndDate = useCallback(
+    (date?: Date) => {
+      if (!date) {
+        return;
+      }
+      endDateInput.handleSetDate(date);
+      setMonth(date);
+    },
+    [endDateInput]
+  );
+
+  /**
+   * Iterate over the dates between a start and end date.
+   */
+  const iterateDateRange = useCallback(
+    (startDate: Date, endDate: Date, cb: (date: Date) => void) => {
+      const normalizedEndDate = endDate || startDate;
+      if (
+        isSameDay(startDate, normalizedEndDate) ||
+        isAfter(startDate, normalizedEndDate)
+      ) {
+        return;
+      }
+
+      if (isSameDay(startDate, endDate)) {
+        return cb(startDate);
+      }
+
+      const days = differenceInDays(normalizedEndDate, startDate);
+      cb(startDate);
+
+      for (let i = 0; i < days; i++) {
+        cb(addDays(startDate, i + 1));
+      }
+    },
+    []
+  );
+
+  /**
+   * Detect if a blocked date is between the start and end date, inclusively,
+   * accounting for the `allow-blocked-end-date` attribute.
+   */
+  const isBlockedRange = useCallback(
+    (startDate?: Date, endDate?: Date) => {
+      if (!startDate || !endDate) {
+        return;
+      }
+
+      let blockedCount = 0;
+      if (!allowBlockedRanges) {
+        iterateDateRange(startDate, endDate, (date) => {
+          if (blockedDates.contains(date)) {
+            blockedCount++;
+          }
+        });
```

Comment:
it seems we only care if blockedCount is greater than 2, I wonder if we can refactor this to stop iterating once we see it is 2 or larger?